### ISSUE TITLE
tf2onnx support dynamic inputs length？

### ISSUE THREAD

**Issue by jianqianzhou** [onnx/tensorflow-onnx#1283]
tf2onnx                  1.8.1
tensorflow               2.3.0
transformers             4.2.1

from transformers import TFBertForSequenceClassification
model = TFBertForSequenceClassification.from_pretrained('./bert-base-chinese')

from transformers import BertTokenizer, glue_convert_examples_to_features
import tensorflow as tf
import tensorflow_datasets as tfds
tokenizer = BertTokenizer.from_pretrained('./bert-base-chinese')
data = tfds.load('glue/mrpc')
train_dataset = glue_convert_examples_to_features(data['train'], tokenizer, max_length=128, task='mrpc')
train_dataset = train_dataset.shuffle(100).batch(32).repeat(2)

optimizer = tf.keras.optimizers.Adam(learning_rate=3e-5)
loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(optimizer=optimizer, loss=loss)
model.fit(train_dataset, epochs=1, steps_per_epoch=115)
model.save('./my_mrpc_model/')

then I use "python -m tf2onnx.convert --saved-model ./my_mrpc_model/ --opset 10 --output model.onnx"
I throught Netron open model.onnx , find input length is 5,but in my program i set length is 128,so tf2onnx not support dynamic input length?

**Comment by guschmue** [comment]
By default we use the serving_default signature and ```saved_model show --dir  ./my_mrpc_model --all``` shows:
```
signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['input_ids'] tensor_info:
        dtype: DT_INT32
        shape: (-1, 5)
        name: serving_default_input_ids:0
  The given SavedModel SignatureDef contains the following output(s):
    outputs['logits'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 2)
        name: StatefulPartitionedCall:0
  Method name is: tensorflow/serving/predict
```
which is the same as pops out in the onnx model:
```
 %input_ids:0[INT32, unk__132x5]
```

**Comment by jianqianzhou** [comment]
> By default we use the serving_default signature and `saved_model show --dir ./my_mrpc_model --all` shows:
> 
> ```
> signature_def['serving_default']:
>   The given SavedModel SignatureDef contains the following input(s):
>     inputs['input_ids'] tensor_info:
>         dtype: DT_INT32
>         shape: (-1, 5)
>         name: serving_default_input_ids:0
>   The given SavedModel SignatureDef contains the following output(s):
>     outputs['logits'] tensor_info:
>         dtype: DT_FLOAT
>         shape: (-1, 2)
>         name: StatefulPartitionedCall:0
>   Method name is: tensorflow/serving/predict
> ```
> 
> which is the same as pops out in the onnx model:
> 
> ```
>  %input_ids:0[INT32, unk__132x5]
> ```

so if I want to modify the input_ids length(from 5 to 128)，What I can do?thans!

**Comment by jianqianzhou** [comment]
> By default we use the serving_default signature and `saved_model show --dir ./my_mrpc_model --all` shows:
> 
> ```
> signature_def['serving_default']:
>   The given SavedModel SignatureDef contains the following input(s):
>     inputs['input_ids'] tensor_info:
>         dtype: DT_INT32
>         shape: (-1, 5)
>         name: serving_default_input_ids:0
>   The given SavedModel SignatureDef contains the following output(s):
>     outputs['logits'] tensor_info:
>         dtype: DT_FLOAT
>         shape: (-1, 2)
>         name: StatefulPartitionedCall:0
>   Method name is: tensorflow/serving/predict
> ```
> 
> which is the same as pops out in the onnx model:
> 
> ```
>  %input_ids:0[INT32, unk__132x5]
> ```

from transformers import TFBertForSequenceClassification
from transformers import BertTokenizer, glue_convert_examples_to_features
import tensorflow as tf
import tensorflow_datasets as tfds
tokenizer = BertTokenizer.from_pretrained('./bert-base-chinese')

model = TFBertForSequenceClassification.from_pretrained('./bert-base-chinese')

data = tfds.load('glue/mrpc')
train_dataset = glue_convert_examples_to_features(data['train'], tokenizer, max_length=128, task='mrpc')
train_dataset = train_dataset.shuffle(100).batch(32).repeat(2)

optimizer = tf.keras.optimizers.Adam(learning_rate=3e-5)
loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(optimizer=optimizer, loss=loss)
model.fit(train_dataset, epochs=1, steps_per_epoch=115)

@tf.function()
def my_predict(input_ids, token_type_ids, attention_mask):
   inputs = {
        'input_ids': input_ids,
        'token_type_ids': token_type_ids,
        'attention_mask': attention_mask
   }
   prediction = model(inputs)[0]
   return {"predict": prediction}

my_signatures = my_predict.get_concrete_function(
     input_ids=tf.TensorSpec([None,None], dtype=tf.dtypes.int32, name="input_ids"),
     token_type_ids=tf.TensorSpec([None,None], dtype=tf.dtypes.int32, name="token_type_ids"),
     attention_mask=tf.TensorSpec([None,None], dtype=tf.dtypes.int32, name="attention_mask"),
)

export_path="./my_mrpc_model/"
# Save the model.
tf.saved_model.save(
    model,
    export_dir=export_path,
    signatures=my_signatures
)
I use this code to modify input and output, and use "python -m tf2onnx.convert --saved-model ./my_mrpc_model/ --opset 11 --output model.onnx" 
I open model.onnx in Netron , I found an error,in onnx format,the input name modify from "input_ids" to "input_ids:0",like this:
![image](https://user-images.githubusercontent.com/62974595/105471834-247f2080-5cd6-11eb-8448-ba597f9a4337.png)
while I convert pytorch to onnx like this is correct
![image](https://user-images.githubusercontent.com/62974595/105471978-555f5580-5cd6-11eb-8eb9-b431fc0b74e4.png)
so  how to change the input from "input_ids:0" to "input_ids"? thanks

**Comment by guschmue** [comment]
Use use the tensor names and not the node names as inputs and outputs the same way native tensorflow would do.
The reason for this is for example if the last op has multiple outputs (topk as example) we need to distinguish between them and tensorflow does this with ":0", ":1" ....
We can maybe think about a command line option to strip the :0 if there is only 1.

**Comment by jianqianzhou** [comment]
> We use the tensor names and not the node names as inputs and outputs the same way native tensorflow would do.
> The reason for this is for example if the last op has multiple outputs (topk as example) we need to distinguish between them and tensorflow does this with ":0", ":1" ....
> Its easy to understand for the outputs. For inputs we use the tensor names as well do be consistent that we only use tensor names and never the node name.
> We can maybe think about a command line option to strip the :0 if there is only 1.

because onnx runtime optimized bert onnx by node names，so if the node names is "input_ids:0" not "input_ids", onnx runtime will miss this，and won't optimize onnx.

**Comment by guschmue** [comment]
got it, let me see if we can add a option. For the onnxruntime optimization - it has some other issues with tensorflow models that it doesn't detect all the fusions it should do but somebody is working on this.

**Comment by TomWildenhain-Microsoft** [comment]
As a workaround, you can use the model editing tool onnx2py and find/replace "input_ids:0" with "input_ids".

https://github.com/microsoft/onnxconverter-common/blob/master/onnxconverter_common/onnx2py.py

**Comment by TomWildenhain-Microsoft** [comment]
You can now convert with the --rename-inputs flag to specify the desired input names.
